### PR TITLE
Unify types for StorageNode, Tracker and SmartContract configs.

### PR DESCRIPTION
**Pull request by timoxley**
This is some low priority typing/refactoring.

### Problem

Currently there's multiple definitions + names for tracker & storage node records.

* `SmartContractRecord` (protocol) is identical to `TrackerRegistryItem` (broker).
* `StorageNodeInfo` (protocol) is identical to `StorageNodeRegistryItem` (broker).
* `SmartContractRecord` is confusingly named as we have multiple smart contracts.

Likely this duplication is because of the way `Protocol` is exported from `network` seems to lose type exports, so they needed to be redefined. Going forward it's possible we should just make protocol a direct dependency of broker.

### Fixes

This is a subset of some changes I've been working with on the `NET-201-brubeck-client` branch.

* `TrackerRegistryItem` and `SmartContractRecord` are now `TrackerRecord` from protocol.
* `StorageNodeRegistryItem` and `StorageNodeInfo` are now `StorageNodeRecord` from protocol.
* Removed the `T extends TrackerInfo` generic parameter from `TrackerRegistry` as it was unused (was always `TrackerInfo` or equivalent). 
* Added a new file `network/src/streamr-client-protocol.ts` that allows you to import the protocol via the network and use it as if it were just the protocol module. e.g.
```typescript
import { StorageNodeRecord } from 'streamr-network/dist/streamr-client-protocol'
```
* Added `getTrackerRecord()` method to the `Tracker` class. This gives the correct shape for consumption of a `tracker` instance in `NetworkNode` `trackers` config. e.g.
```typescript
publisherNode = createNetworkNode({
    id: 'publisherNode',
    trackers: [tracker.getTrackerRecord()]
})
```
* `{ contractAddress: string, jsonRpcProvider: string | ConnectionInfo }` is now `SmartContractConfig` from protocol. This also addresses some inconsistency with the typing of `jsonRpcProvider` as `string | ConnectionInfo` vs just `string`. 
* Also updated the broker config JSON schema to be consistent with this change using the typescript definition of `ConnectionInfo` (via `typescript-json-schema`).

### CODE REVIEW

File: packages/network/src/identifiers.ts [hunk 1]
```diff
@@ -1,7 +1,9 @@
-import { TrackerLayer, SmartContractRecord } from 'streamr-client-protocol'
+import { TrackerLayer } from 'streamr-client-protocol'
 import { NodeId } from './logic/Node'
 import { RtcSubTypes } from './logic/RtcMessage'
 
+export { TrackerRecord } from 'streamr-client-protocol'
```
Comment (juslesan) [inline packages/network/src/identifiers.ts:5]: Would it make more sense to import TrackerRecord directly from streamr-client-protocol to modules with these changes?

File: packages/network/src/streamr-client-protocol.ts [hunk 2]
```diff
@@ -0,0 +1,3 @@
+import Protocol from 'streamr-client-protocol'
+export * from 'streamr-client-protocol'
+export default Protocol
```
Comment (juslesan) [inline packages/network/src/streamr-client-protocol.ts:3]: I feel like there is a need to get rid of the instanceof checking for messages in the protocol to avoid having to strictly use the network's protocol instance in the broker and elsewhere.

**Review by juslesan — APPROVED** [review]

File: packages/broker/src/helpers/config.schema.json [hunk 3]
```diff
@@ -1,304 +1,350 @@
 {
```
Comment (harbu) [inline packages/broker/src/helpers/config.schema.json:1]: Why these changes? Are these just indentation?

File: packages/network/src/identifiers.ts (hunk 1 above)
Comment (harbu) [reply packages/network/src/identifiers.ts:5]: Yeah re-exporting seems like needless indirection? Just import from protocol directly where needed?

File: packages/network/src/streamr-client-protocol.ts [hunk 4]
```diff
@@ -0,0 +1,3 @@
+import Protocol from 'streamr-client-protocol'
```
Comment (harbu) [inline packages/network/src/streamr-client-protocol.ts:1]: Maybe the contents of this file should live in `composition.ts` so instead of
```ts
import { ... } from 'streamr-network/dist/streamr-client-protocol'
```

We can just do 
```ts
import { ... } from 'streamr-network'
```

(Which I find better because the latter is relying on purposefully exported things instead of referring to an arbitrary file within the project.)

We could also consider just adding `streamr-client-protocol` as a dependency of broker and using stuff from there directly, which would be arguably less confusing. Why we haven't done so before? Becuase of concerns of potential version mismatches between network and broker which could lead to weird behaviour.

**Review by harbu — COMMENTED** [review]
Good stuff; see my comments on import

File: packages/broker/src/config.ts [hunk 5]
```diff
@@ -1,13 +1,4 @@
...
-    contractAddress: string
-    jsonRpcProvider: string
-}
-
-export interface TrackerRegistryItem {
-    id: string
-    ws: string
-    http: string
-}
+import { SmartContractConfig, TrackerRecord, StorageNodeRecord } from 'streamr-network/dist/streamr-client-protocol'
```
Comment (teogeb) [inline packages/broker/src/config.ts:1]: Maybe we could avoid importing the internals of `streams-network`? E.g. 
```
import { Protocol } from 'streamr-network'
trackers: Protocol.TrackerRecord[] | Protocol.SmartContractConfig,
```

File: packages/protocol/src/utils/TrackerRegistry.ts [hunk 6]
```diff
@@ -37,40 +31,27 @@ export class TrackerRegistry<T extends TrackerInfo> {
...
-
-    const contract = new Contract(contractAddress, trackerRegistryConfig.abi, provider)
-    // check that contract is connected
-    await contract.addressPromise
-
+async function fetchTrackers(config: SmartContractConfig) {
+    const contract = await initContract(config, trackerRegistryConfig.abi)
     if (typeof contract.getNodes !== 'function') {
-        throw Error(`getNodes function is not defined in smart contract (${contractAddress})`)
+        throw Error(`getNodes function is not defined in smart contract (${config.contractAddress})`)
```
Comment (teogeb) [inline packages/protocol/src/utils/TrackerRegistry.ts:42]: It seems that the checking of available functions is a typical use case. Maybe `initContract` could implement this checking if we'd add a third parameter which lists all required functions?

File: packages/protocol/src/utils/StorageNodeRegistry.ts [hunk 7]
```diff
@@ -1,20 +1,16 @@
...
+import { initContract, SmartContractConfig } from './SmartContractUtil'
 
 // storageNodeAddress => HTTP address
-export type StorageNodeInfo = {
+export type StorageNodeRecord = {
     address: string
     url: string
 }
 
 export class StorageNodeRegistry {
```
Comment (teogeb) [inline packages/protocol/src/utils/StorageNodeRegistry.ts:10]: Maybe we could move this class to `Broker` package as storage functionality is nowadays considered to be an implementation detail of a broker node, and not protocol level logic.

(At some point we should update the https://github.com/streamr-dev/streamr-specs/blob/master/PROTOCOL.md#storagenodesrequest spec and officially deprecate `StorageNodesRequest` and `StorageNodesRequest`)

There is already a class `network-monorepo/packages/broker/src/StorageNodeRegistry.ts`. This functionality could be merged to that file.

File: packages/network/src/logic/Tracker.ts [hunk 8]
```diff
@@ -254,4 +254,12 @@ export class Tracker extends EventEmitter {
     getOverlayPerStream(): Readonly<OverlayPerStream> {
         return this.overlayPerStream
     }
+
+    getTrackerRecord(): TrackerRecord {
+        return {
+            id: this.peerInfo.peerId,
+            http: this.getUrl().replace(/^ws/, 'http'),
```
Comment (teogeb) [inline packages/network/src/logic/Tracker.ts:261]: If the tracker doesn't have HTTP endpoints, this field should be `undefined`? See https://github.com/streamr-dev/network-monorepo/blob/06137024a0e56bd24e601fe27ac6811d0ffe3dc6/packages/network/src/composition.ts#L90

Currently the `TrackerRecord` type definition doesn't allow `undefined` field value. Does the smart contract allow it?

File: packages/protocol/src/utils/StorageNodeRegistry.ts (hunk 7 above)
Comment (samt1803) [reply packages/protocol/src/utils/StorageNodeRegistry.ts:10]: In the decentralised version I am working on, the StorageNodeRegistry is residing (like the StreamRegistry) in the client and the one in the broker is being replaced by just calling the client. So this StorageNodeRecord is needed in the client for CRUD operations on storagenodes or adding streams to storagenodes. So it could be moved to the client, and then the broker uses the definition from the client instead of the protocol..

**Review by samt1803 — APPROVED** [review]

File: packages/broker/src/helpers/config.schema.json (hunk 3 above)
Comment (timoxley) [reply packages/broker/src/helpers/config.schema.json:1]: Ah must have autoformatted the whitespace. It's now 4 space indentation like rest of code.

File: packages/protocol/src/utils/TrackerRegistry.ts (hunk 6 above)
Comment (timoxley) [reply packages/protocol/src/utils/TrackerRegistry.ts:42]: Good thought, but let's add it if/when there's more than just these two.

File: packages/network/src/logic/Tracker.ts (hunk 8 above)
Comment (timoxley) [reply packages/network/src/logic/Tracker.ts:261]: > if the tracker doesn't have HTTP endpoints

Good point

File: packages/network/src/streamr-client-protocol.ts (hunk 4 above)
Comment (timoxley) [reply packages/network/src/streamr-client-protocol.ts:1]: > concerns of potential version mismatches between network and broker which could lead to weird behaviour

Yeah I figured there might be some issues with module-level/class static state or `instanceof` and didn't want to audit every file so just left that part as-is.

Comment (timoxley) [reply packages/network/src/streamr-client-protocol.ts:1]: Though I guess there may be similar issues with client's internal protocol vs broker + network protocol, either versioning or simply a different instance.

Comment (timoxley) [reply packages/network/src/streamr-client-protocol.ts:1]: related https://github.com/streamr-dev/network-monorepo/pull/208#discussion_r714559277

File: packages/network/src/logic/Tracker.ts (hunk 8 above)
Comment (timoxley) [reply packages/network/src/logic/Tracker.ts:261]: > Does the smart contract allow it?

Supports anything, the data is a json blob as a string.

**Comment by harbu** [comment]
Status of this, will be resolving conflicts or close PR? @timoxley
